testsuite/smoke-test/src/randomness/validator_restart_during_dkg.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    randomness::{
+        decrypt_key_map, get_on_chain_resource, verify_dkg_transcript, wait_for_dkg_finish,
+    },
+    smoke_test_environment::SwarmBuilder,
+};
+use aptos_forge::{NodeExt, SwarmExt};
+use aptos_logger::{debug, info};
+use aptos_rest_client::Client;
+use aptos_types::{
+    dkg::DKGState,
+    on_chain_config::{FeatureFlag, Features},
+};
+use futures::future::join_all;
+use std::{sync::Arc, time::Duration};
+
+#[tokio::test]
+async fn validator_restart_during_dkg() {
+    let epoch_duration_secs = 30;
+    let estimated_dkg_latency_secs = 30;
+    let time_limit_secs = epoch_duration_secs + estimated_dkg_latency_secs;
+    let num_validators = 4;
+    let num_validators_to_restart = 3;
+    let mut swarm = SwarmBuilder::new_local(num_validators)
+        .with_num_fullnodes(1)
+        .with_aptos()
+        .with_init_config(Arc::new(|_, conf, _| {
+            conf.api.failpoints_enabled = true;
+        }))
+        .with_init_genesis_config(Arc::new(|conf| {
+            conf.epoch_duration_secs = 30;
+
+            // Ensure vtxn is enabled.
+            conf.consensus_config.enable_validator_txns();
+
+            // Ensure randomness flag is set.
+            let mut features = Features::default();
+            features.enable(FeatureFlag::RECONFIGURE_WITH_DKG);
+            conf.initial_features_override = Some(features);
+        }))
+        .build()
+        .await;
+
+    swarm
+        .wait_for_all_nodes_to_catchup_to_epoch(2, Duration::from_secs(epoch_duration_secs * 10))
+        .await
+        .unwrap();
+
+    let decrypt_key_map = decrypt_key_map(&swarm);
+
+    info!("Wait for an epoch start.");
+    let validator_clients: Vec<Client> =
+        swarm.validators().map(|node| node.rest_client()).collect();
+    let dkg_session_1 = wait_for_dkg_finish(&validator_clients[3], None, time_limit_secs).await;
+
+    info!(
+        "Current epoch is {}.",
+        dkg_session_1.metadata.dealer_epoch + 1
+    );
+
+    info!("Inject fault to all validators so they get stuck upon the first DKG message received.");
```

Comment:
Which epoch do the validators crash? If it is any epoch other than the first epoch, this test also tested the crash recovery of randomness manager right?

Reply:
we have `wait_for_all_nodes_to_catchup_to_epoch(2, ..)` so yes, also kinda covered rand_manager crash recovery.